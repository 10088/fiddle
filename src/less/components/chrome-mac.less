.darwin {
  .chrome {
    text-align: center;
    background-color: @background-1;
    border-bottom: @border;
    -webkit-app-region: drag;
    height: 23px;
    font-size: 1.4rem;

    small,
    p,
    span {
      user-select: none;
    }
  }
}
