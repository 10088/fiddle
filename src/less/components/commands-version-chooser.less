button.version-chooser.bp3-button {
  border-radius: 3px;
}
