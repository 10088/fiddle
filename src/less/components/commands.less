@import "../variables.less";

header {
  box-shadow: 0 0 0 1px rgba(16, 22, 26, 0.2), 0 0 0 rgba(16, 22, 26, 0), 0 1px 1px rgba(16, 22, 26, 0.4);
  background-color: @background-1;
  -webkit-app-region: drag;
}

.commands {
  display: flex;
  justify-content: space-between;
  height: 50px;
  padding: 10px;
  box-sizing: border-box;
  font-family: @fonts-common;

  button,
  input,
  select {
    -webkit-app-region: no-drag;
    margin-bottom: 0;
  }

  select {
    width: 100px;
    margin-left: 5px;
  }

  & > div {
    display: flex;

    &:last-of-type {
      margin-right: 10px;

      button {
        margin-left: 5px;
      }
    }
  }

  .address-bar {
    height: 28px;
    width: 360px;
    transition: width 0.4s;
    display: flex;

    input,
    span {
      font-family: @fonts-common;
      font-weight: 300;
      font-size: 11px;
      cursor: auto;
      -webkit-app-region: no-drag;
    }

    input {
      flex-grow: 1;
      background: white;
      padding-left: 1rem;
      height: 2.6rem;
    }

    input:invalid {
      box-shadow: inset 0 0 0 1px #e4252d;
    }

    span {
      background: white;
      color: #1d1d1d;
      border-bottom-color: rgb(209, 209, 209);
      border-bottom-style: solid;
      border-bottom-left-radius: 4px;
      border-bottom-width: 1px;
      border-image-outset: 0;
      border-image-repeat: stretch;
      border-image-slice: 100%;
      border-image-source: none;
      border-image-width: 1;
      border-left-color: rgb(209, 209, 209);
      border-left-style: solid;
      border-left-width: 1px;
      border-top-color: rgb(209, 209, 209);
      border-top-left-radius: 4px;
      border-top-style: solid;
      border-top-width: 1px;
      display: inline-block;
      height: 1.2rem;
      letter-spacing: normal;
      margin: 0;
      padding: 6px 0 10px 10px;
      user-select: text;
      width: 132px;
      text-rendering: auto;
      text-shadow: none;
      text-transform: none;
    }

    &.empty {
      width: 200px;
    }

    &:focus-within {
      span,
      input {
        border-color: @foreground-2;
      }
    }
  }
}
