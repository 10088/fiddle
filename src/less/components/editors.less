.mosaic-tile {
  animation: fadein 0.15s;
}

.mosaic-window-toolbar  {
  & > div {
    display: flex;
    justify-content: space-between;
    width: 100%;
    padding: 0 3px 0 29px;

    .mosaic-controls {
      transform: scale(0.7);
    }

    h5 {
      margin: 4px 0 0 0;
    }

    button {
      margin-left: 8px;
    }
  }
}

@keyframes fadein {
  from { opacity: 0; }
  to { opacity: 1; }
}
