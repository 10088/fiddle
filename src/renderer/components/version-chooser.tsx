import * as React from 'react';
import * as semver from 'semver';
import { observer } from 'mobx-react';

import { AppState } from '../state';

export interface VersionChooserState {
  value: string;
}

export interface VersionChooserProps {
  appState: AppState;
}

/**
 * A dropdown allowing the selection of Electron versions. The actual
 * download is managed in the state.
 *
 * @class VersionChooser
 * @extends {React.Component<VersionChooserProps, VersionChooserState>}
 */
@observer
export class VersionChooser extends React.Component<VersionChooserProps, VersionChooserState> {
  constructor(props: VersionChooserProps) {
    super(props);

    this.handleChange = this.handleChange.bind(this);
  }

  /**
   * Handle change, which usually means that we'd like update
   * the selection version.
   *
   * @param {React.ChangeEvent<HTMLSelectElement>} event
   */
  public handleChange(event: React.ChangeEvent<HTMLSelectElement>) {
    this.props.appState.setVersion(event.target.value);
  }

  /**
   * Renders the individual options (Electron versions)
   *
   * @returns {Array<JSX.Element>}
   */
  public renderOptions(): Array<JSX.Element> {
    const { versions } = this.props.appState;

    return Object.keys(versions)
      .sort((a, b) => semver.gt(a, b, true) ? -1 : 1)
      .map((key) => {
        const { tag_name, state } = versions[key];
        const version = tag_name;
        const icon = state === 'ready'
          ? '✅'
          : state === 'downloading' ? '⏬' : '⏹';

        return (
          <option value={version} key={version}>
            {icon} {version}
          </option>
        );
      });
  }

  public render() {
    return (
      <select value={`v${this.props.appState.version}`} onChange={this.handleChange}>
        {this.renderOptions()}
      </select>
    );
  }
}
