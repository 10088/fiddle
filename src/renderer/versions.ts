import { GitHubVersion } from '../interfaces';

/**
 * Retrieves our best guess regarding the latest Electron versions. Tries to
 * fetch them from localStorage, then from a static releases.json file.
 *
 * @returns {Array<GitHubVersion>}
 */
export function getKnownVersions(): Array<GitHubVersion> {
  const fromLs = window.localStorage.getItem('known-electron-versions');

  if (fromLs) {
    try {
      return JSON.parse(fromLs);
    } catch (error) {
      console.warn(`Parsing known versions failed, falling back`);
    }
  }

  return require('../../static/releases.json');
}

/**
 * Saves known versions to localStorage.
 *
 * @param {Array<GitHubVersion>} versions
 */
export function saveKnownVersions(versions: Array<GitHubVersion>) {
  const stringified = JSON.stringify(versions);
  window.localStorage.setItem('known-electron-versions', stringified);
}

/**
 * Tries to refresh our known versions and returns whatever we have
 * saved after.
 *
 * @export
 * @returns {Promise<Array<GitHubVersion>>}
 */
export async function getUpdatedKnownVersions(
  pages: number
): Promise<Array<GitHubVersion>> {
  try {
    await fetchVersions(pages);
  } catch (error) {
    console.warn(`Versions: Failed to fetch versions`, { error });
  }

  return getKnownVersions();
}

/**
 * Fetch the latest known versions directly from GitHub.
 *
 * @returns {Promise<Array<GitHubVersion>>}
 */
export async function fetchVersions(pages: number) {
  const output: Array<GitHubVersion> = [];

  for (let i = 0; i < pages; i++) {
    const page = await fetchVersionPage(i + 1);

    if (page && page.length > 0 && page[0].tag_name) {
      page.forEach((release) => {
        const updated = {
          url: release.url,
          assets_url: release.assets_url,
          html_url: release.html_url,
          tag_name: release.tag_name,
          target_commitish: release.target_commitish,
          name: release.name,
          prerelease: release.prerelease,
          created_at: release.created_at,
          published_at: release.published_at,
          body: release.body
        };

        output.push(updated as GitHubVersion);
      });
    }
  }

  saveKnownVersions(output);
}

/**
 * Fetch a single releases page from GitHub.
 *
 * @returns {Promise<Array<GitHubVersion>>}
 */
export function fetchVersionPage(page?: number): Promise<Array<GitHubVersion>> {
  const url = `https://api.github.com/repos/electron/electron/releases`
    + ((page && page > 1) ? `?page=${page}` : '');

  return window.fetch(url)
    .then((response) => response.json());
}
