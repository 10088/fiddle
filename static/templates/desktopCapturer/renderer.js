const { desktopCapturer } = require('electron')

// The following example shows how to capture video from
// the screen. It also grabs each window, so you could
// just grab video from a single window.
//
desktopCapturer.getSources({ types: ['window', 'screen'] }, (error, sources) => {
  if (error) throw error
  for (let i = 0; i < sources.length; ++i) {
    console.log(sources[i])

    if (sources[i].name.startsWith('Screen')) {
      navigator.mediaDevices.getUserMedia({
        audio: false,
        video: {
          mandatory: {
            chromeMediaSource: 'desktop',
            chromeMediaSourceId: sources[i].id,
            minWidth: 1280,
            maxWidth: 1280,
            minHeight: 720,
            maxHeight: 720
          }
        }
      })
      .then((stream) => handleStream(stream))
      .catch((e) => handleError(e))
      return
    }
  }
})

function handleStream (stream) {
  const video = document.querySelector('video')
  video.srcObject = stream
  video.onloadedmetadata = (e) => video.play()
}

function handleError (e) {
  console.log(e)
}