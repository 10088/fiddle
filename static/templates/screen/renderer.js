const { screen } = require('electron')

const screens = screen.getAllDisplays()
const div = document.querySelector('div')

div.innerText = JSON.stringify(screens);
