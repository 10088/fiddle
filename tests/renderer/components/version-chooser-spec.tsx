import { shallow } from 'enzyme';
import * as React from 'react';

import { VersionChooser } from '../../../src/renderer/components/version-chooser';
import { ElectronReleaseChannel } from '../../../src/renderer/versions';
import { mockVersions } from '../../mocks/electron-versions';

describe('VersionChooser component', () => {
  beforeEach(() => {
    this.store = {
      versions: mockVersions,
      versionsToShow: [ ElectronReleaseChannel.stable, ElectronReleaseChannel.beta ],
      setVersion: jest.fn()
    };
  });

  it('renders', () => {
    const wrapper = shallow(<VersionChooser appState={this.store} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('handles a change appropriately', () => {
    const wrapper = shallow(<VersionChooser appState={this.store} />);
    wrapper.find('select').simulate('change', { target: { value: 'v2.0.0' } });

    expect(this.store.setVersion).toHaveBeenCalledWith('v2.0.0');
  });
});
