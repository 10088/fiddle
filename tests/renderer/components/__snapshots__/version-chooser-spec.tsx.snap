// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VersionChooser component renders 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <VersionChooser
    appState={
      Object {
        "setVersion": [MockFunction],
        "versions": Object {
          "1.8.7": Object {
            "assets_url": "https://api.github.com/repos/electron/electron/releases/11032343/assets",
            "body": "## Bug Fixes

* Fixed context menu for sandbox devtools. #12734

*...",
            "created_at": "2018-05-16T18:15:18Z",
            "html_url": "https://github.com/electron/electron/releases/tag/v1.8.7",
            "name": "electron v1.8.7",
            "prerelease": false,
            "published_at": "2018-05-16T18:21:33Z",
            "state": "unknown",
            "tag_name": "v1.8.7",
            "target_commitish": "1-8-x",
            "url": "https://api.github.com/repos/electron/electron/releases/11032343",
          },
          "2.0.1": Object {
            "assets_url": "https://api.github.com/repos/electron/electron/releases/11032425/assets",
            "body": "## Bug Fixes

* Fixed flaky security-warnings test. #12776,...",
            "created_at": "2018-05-16T17:30:26Z",
            "html_url": "https://github.com/electron/electron/releases/tag/v2.0.1",
            "name": "electron v2.0.1",
            "prerelease": false,
            "published_at": "2018-05-16T18:40:54Z",
            "state": "unknown",
            "tag_name": "v2.0.1",
            "target_commitish": "2-0-x",
            "url": "https://api.github.com/repos/electron/electron/releases/11032425",
          },
          "2.0.2": Object {
            "assets_url": "https://api.github.com/repos/electron/electron/releases/11120972/assets",
            "body": "## Bug Fixes

* Fixed long jitter buffer delays...",
            "created_at": "2018-05-22T18:52:16Z",
            "html_url": "https://github.com/electron/electron/releases/tag/v2.0.2",
            "name": "electron v2.0.2",
            "prerelease": false,
            "published_at": "2018-05-22T20:14:35Z",
            "state": "unknown",
            "tag_name": "v2.0.2",
            "target_commitish": "2-0-x",
            "url": "https://api.github.com/repos/electron/electron/releases/11120972",
          },
        },
        "versionsToShow": Array [
          "Stable",
          "Beta",
        ],
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <option
          value="v2.0.2"
        >
          ⏹
           
          v2.0.2
        </option>,
        <option
          value="v2.0.1"
        >
          ⏹
           
          v2.0.1
        </option>,
        <option
          value="v1.8.7"
        >
          ⏹
           
          v1.8.7
        </option>,
      ],
      "className": "select-versions",
      "onChange": [Function],
      "value": "vundefined",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": "v2.0.2",
        "nodeType": "host",
        "props": Object {
          "children": Array [
            "⏹",
            " ",
            "v2.0.2",
          ],
          "value": "v2.0.2",
        },
        "ref": null,
        "rendered": Array [
          "⏹",
          " ",
          "v2.0.2",
        ],
        "type": "option",
      },
      Object {
        "instance": null,
        "key": "v2.0.1",
        "nodeType": "host",
        "props": Object {
          "children": Array [
            "⏹",
            " ",
            "v2.0.1",
          ],
          "value": "v2.0.1",
        },
        "ref": null,
        "rendered": Array [
          "⏹",
          " ",
          "v2.0.1",
        ],
        "type": "option",
      },
      Object {
        "instance": null,
        "key": "v1.8.7",
        "nodeType": "host",
        "props": Object {
          "children": Array [
            "⏹",
            " ",
            "v1.8.7",
          ],
          "value": "v1.8.7",
        },
        "ref": null,
        "rendered": Array [
          "⏹",
          " ",
          "v1.8.7",
        ],
        "type": "option",
      },
    ],
    "type": "select",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <option
            value="v2.0.2"
          >
            ⏹
             
            v2.0.2
          </option>,
          <option
            value="v2.0.1"
          >
            ⏹
             
            v2.0.1
          </option>,
          <option
            value="v1.8.7"
          >
            ⏹
             
            v1.8.7
          </option>,
        ],
        "className": "select-versions",
        "onChange": [Function],
        "value": "vundefined",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": "v2.0.2",
          "nodeType": "host",
          "props": Object {
            "children": Array [
              "⏹",
              " ",
              "v2.0.2",
            ],
            "value": "v2.0.2",
          },
          "ref": null,
          "rendered": Array [
            "⏹",
            " ",
            "v2.0.2",
          ],
          "type": "option",
        },
        Object {
          "instance": null,
          "key": "v2.0.1",
          "nodeType": "host",
          "props": Object {
            "children": Array [
              "⏹",
              " ",
              "v2.0.1",
            ],
            "value": "v2.0.1",
          },
          "ref": null,
          "rendered": Array [
            "⏹",
            " ",
            "v2.0.1",
          ],
          "type": "option",
        },
        Object {
          "instance": null,
          "key": "v1.8.7",
          "nodeType": "host",
          "props": Object {
            "children": Array [
              "⏹",
              " ",
              "v1.8.7",
            ],
            "value": "v1.8.7",
          },
          "ref": null,
          "rendered": Array [
            "⏹",
            " ",
            "v1.8.7",
          ],
          "type": "option",
        },
      ],
      "type": "select",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
